File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/JavaScript/122-Best-Time-to-Buy-and-Sell-Stock-II-Tabulation.js
/**
 * Problem: 122. Best Time to Buy and Sell Stock II
 * Approach: Tabulation
 *
 * dp[idx][buy] represents the maximum profit
 * starting from idx with the given buy state.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} prices
 * @return {number}
 */

var maxProfit = function(prices) {
    const n = prices.length;

    const dp = Array.from(
        { length: n + 1 },
        () => Array(2).fill(0)
    );

    for (let idx = n - 1; idx >= 0; idx--) {

        dp[idx][1] = Math.max(
            -prices[idx] + dp[idx + 1][0],
            dp[idx + 1][1]
        );

        dp[idx][0] = Math.max(
            prices[idx] + dp[idx + 1][1],
            dp[idx + 1][0]
        );
    }

    return dp[0][1];
};

File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/Python/122-Best-Time-to-Buy-and-Sell-Stock-II-Tabulation.py
"""
Problem: 122. Best Time to Buy and Sell Stock II
Approach: Tabulation

Time Complexity: O(n)
Space Complexity: O(n)
"""


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        n = len(prices)

        dp = [[0] * 2 for _ in range(n + 1)]

        for idx in range(n - 1, -1, -1):
            dp[idx][1] = max(
                -prices[idx] + dp[idx + 1][0],
                dp[idx + 1][1]
            )

            dp[idx][0] = max(
                prices[idx] + dp[idx + 1][1],
                dp[idx + 1][0]
            )

        return dp[0][1]